### PR TITLE
8333265: De-duplicate method references in java.util.stream.FindOps

### PR DESCRIPTION
**Pull request by cl4es**
Extracting duplicate method references to static field reduce proxy class spinning and loading. In this case 2 less classes loaded when using `findAny()` on each type of stream.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333265](https://bugs.openjdk.org/browse/JDK-8333265): De-duplicate method references in java.util.stream.FindOps (**Enhancement** - P5)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19477/head:pull/19477` \
`$ git checkout pull/19477`

Update a local copy of the PR: \
`$ git checkout pull/19477` \
`$ git pull https://git.openjdk.org/jdk.git pull/19477/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19477`

View PR using the GUI difftool: \
`$ git pr show -t 19477`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19477.diff">https://git.openjdk.org/jdk/pull/19477.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19477#issuecomment-2139497512)

### CODE REVIEW

**Comment by cl4es** [comment]
> Should we extract them to private static utility methods for potential laziness support in the future?

Ideally we shouldn't have to do any of this. I thought such method references were already de-duplicated in javac - at least within the same compilation units - but I saw this in a startup profile and was surprised myself that the demonstrated manual de-duplication has an observable effect.

**Comment by liach** [comment]
Indeed, CallSites are created per indy instruction instead of per CP indy entry (required by JVMS); thus sharing instruction either in initializers or static methods would have the same effect. javac only deduplicates the CP entry.

**Comment by forax** [comment]
For constant method reference, the solution is to use a constant dynamic instead of an invokedynamic.

**Comment by cl4es** [comment]
Vicente filed a bug on javac to investigate this: https://bugs.openjdk.org/browse/JDK-8333278

I wouldn't mind using condy for constant aka non-capturing lambdas. I recall we had a prototype for that years ago, but switching over was shelved for some reason.

File: src/java.base/share/classes/java/util/stream/FindOps.java [hunk 1]
```diff
@@ -194,13 +194,14 @@ public Optional<T> get() {
...
-                    StreamShape.REFERENCE, Optional.empty(),
-                    Optional::isPresent, FindSink.OfRef::new);
+            private static final Predicate<Optional<Object>> IS_PRESENT = Optional::isPresent;
+            private static final Supplier<TerminalSink<Object, Optional<Object>>> NEW
+                    = FindOps.FindSink.OfRef::new;
+            static final FindOp<?, ?> OP_FIND_FIRST = new FindOp<>(true,
+                    StreamShape.REFERENCE, Optional.empty(), IS_PRESENT, NEW);
+
+            static final FindOp<?, ?> OP_FIND_ANY = new FindOp<>(false,
+                    StreamShape.REFERENCE, Optional.empty(), IS_PRESENT, NEW);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/stream/FindOps.java:204]:
```suggestion
            static final TerminalOp<?, ?> OP_FIND_FIRST, OP_FIND_ANY;
            static {
                Predicate<Optional<Object>> isPresent = Optional::isPresent;
                Supplier<TerminalSink<Object, Optional<Object>>> newSink
                    = FindSink.OfRef::new;

                OP_FIND_FIRST = new FindOp<>(true, StreamShape.REFERENCE,
                    Optional.empty(), isPresent, newSink);

                OP_FIND_ANY = new FindOp<>(false, StreamShape.REFERENCE,
                    Optional.empty(), isPresent, newSink);
            }
```

File: src/java.base/share/classes/java/util/stream/FindOps.java [hunk 2]
```diff
@@ -217,12 +218,13 @@ public OptionalInt get() {
...
+            private static final Supplier<TerminalSink<Integer, OptionalInt>> NEW
+                    = FindOps.FindSink.OfInt::new;
             static final TerminalOp<Integer, OptionalInt> OP_FIND_FIRST = new FindOp<>(true,
-                    StreamShape.INT_VALUE, OptionalInt.empty(),
-                    OptionalInt::isPresent, FindSink.OfInt::new);
+                    StreamShape.INT_VALUE, OptionalInt.empty(), IS_PRESENT, NEW);
             static final TerminalOp<Integer, OptionalInt> OP_FIND_ANY = new FindOp<>(false,
-                    StreamShape.INT_VALUE, OptionalInt.empty(),
-                    OptionalInt::isPresent, FindSink.OfInt::new);
+                    StreamShape.INT_VALUE, OptionalInt.empty(), IS_PRESENT, NEW);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/stream/FindOps.java:227]:
```suggestion
            static final TerminalOp<Integer, OptionalInt> OP_FIND_FIRST, OP_FIND_ANY;
            static {
                Predicate<OptionalInt> isPresent = OptionalInt::isPresent;
                Supplier<TerminalSink<Integer, OptionalInt> newSink
                    = FindSink.OfInt::new;

                OP_FIND_FIRST = new FindOp<>(true, StreamShape.INT_VALUE,
                    OptionalInt.empty(), isPresent, newSink);

                OP_FIND_ANY = new FindOp<>(false, StreamShape.INT_VALUE,
                    OptionalInt.empty(), isPresent, newSink);
            }
```

File: src/java.base/share/classes/java/util/stream/FindOps.java [hunk 3]
```diff
@@ -239,12 +241,13 @@ public OptionalLong get() {
...
+            private static final Supplier<TerminalSink<Long, OptionalLong>> NEW
+                    = FindOps.FindSink.OfLong::new;
             static final TerminalOp<Long, OptionalLong> OP_FIND_FIRST = new FindOp<>(true,
-                    StreamShape.LONG_VALUE, OptionalLong.empty(),
-                    OptionalLong::isPresent, FindSink.OfLong::new);
+                    StreamShape.LONG_VALUE, OptionalLong.empty(), IS_PRESENT, NEW);
             static final TerminalOp<Long, OptionalLong> OP_FIND_ANY = new FindOp<>(false,
-                    StreamShape.LONG_VALUE, OptionalLong.empty(),
-                    OptionalLong::isPresent, FindSink.OfLong::new);
+                    StreamShape.LONG_VALUE, OptionalLong.empty(), IS_PRESENT, NEW);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/stream/FindOps.java:250]:
```suggestion
            static final TerminalOp<Long, OptionalLong> OP_FIND_FIRST, OP_FIND_ANY;
            static {
                Predicate<OptionalLong> isPresent = OptionalLong::isPresent;
                Supplier<TerminalSink<Long, OptionalLong> newSink
                    = FindSink.OfLong::new;

                OP_FIND_FIRST = new FindOp<>(true, StreamShape.LONG_VALUE,
                    OptionalLong.empty(), isPresent, newSink);

                OP_FIND_ANY = new FindOp<>(false, StreamShape.LONG_VALUE,
                    OptionalLong.empty(), isPresent, newSink);
            }
```

File: src/java.base/share/classes/java/util/stream/FindOps.java [hunk 4]
```diff
@@ -261,12 +264,13 @@ public OptionalDouble get() {
...
+            private static final Supplier<TerminalSink<Double, OptionalDouble>> NEW
+                    = FindOps.FindSink.OfDouble::new;
             static final TerminalOp<Double, OptionalDouble> OP_FIND_FIRST = new FindOp<>(true,
-                    StreamShape.DOUBLE_VALUE, OptionalDouble.empty(),
-                    OptionalDouble::isPresent, FindSink.OfDouble::new);
+                    StreamShape.DOUBLE_VALUE, OptionalDouble.empty(), IS_PRESENT, NEW);
             static final TerminalOp<Double, OptionalDouble> OP_FIND_ANY = new FindOp<>(false,
-                    StreamShape.DOUBLE_VALUE, OptionalDouble.empty(),
-                    OptionalDouble::isPresent, FindSink.OfDouble::new);
+                    StreamShape.DOUBLE_VALUE, OptionalDouble.empty(), IS_PRESENT, NEW);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/stream/FindOps.java:273]:
```suggestion
            static final TerminalOp<Double, OptionalDouble> OP_FIND_FIRST, OP_FIND_ANY;
            static {
                Predicate<OptionalDouble> isPresent = OptionalDouble::isPresent;
                Supplier<TerminalSink<Double, OptionalLong> newSink
                    = FindSink.OfDouble::new;

                OP_FIND_FIRST = new FindOp<>(true, StreamShape.DOUBLE_VALUE,
                    OptionalDouble.empty(), isPresent, newSink);

                OP_FIND_ANY = new FindOp<>(false, StreamShape.DOUBLE_VALUE,
                    OptionalDouble.empty(), isPresent, newSink);
            }
```

**Review by ExE-Boss — COMMENTED** [review]
The intermediate `Predicate`s and `Supplier`s don’t need to be fields, and it’s probably better for them to be locals:

**Comment by cl4es** [comment]
Friendly reminder that this improvement needs a Reviewer. While it'd be great if javac can make this redundant this is adding overhead in the meantime, and the temporary workaround doesn't obfuscate the code.

File: test/micro/org/openjdk/bench/java/util/stream/ops/ref/FindAny.java [hunk 5]
```diff
@@ -60,4 +60,10 @@ public Long par_invoke() {
         return LongStream.range(0, size).parallel().boxed().findAny().get();
     }
 
+    public static void main(String... args) {
```
Comment (liach) [inline test/micro/org/openjdk/bench/java/util/stream/ops/ref/FindAny.java:63]: Is this driver necessary?

Comment (liach) [reply test/micro/org/openjdk/bench/java/util/stream/ops/ref/FindAny.java:63]: @cl4es Just wonder what your use case is for this addition. If this is accidentally committed, please remove it and I am glad with all other changes.

Comment (cl4es) [reply test/micro/org/openjdk/bench/java/util/stream/ops/ref/FindAny.java:63]: Adding a `main` method in micros like these allow me to easily multi-purpose them as relatively clean, diagnostic startup microbenchmarks. It's something I've started adding to all micros I author. I guess I should do a write-up about it some time.

Comment (liach) [reply test/micro/org/openjdk/bench/java/util/stream/ops/ref/FindAny.java:63]: Thanks for the explanation.

**Review by liach — APPROVED** [review]

**Pull request closed without merging**